pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@link org.postgresql.core.Parser} stores information about placeholder occurrences in ParameterContext.
+ * In case of standard JDBC placeholders {@code '?'} the position in the SQL text is recorded. For named
+ * placeholders {@code ":paramName"} the name is recorded as well as the position. {@code PgPreparedStatement} can
+ * then use the name to lookup the parameter corresponding index. These values are also used by
+ * toString() methods to provide a human readable representation of the SQL text.
+ */
+public class ParameterContext {
+
+  enum BindStyle {
+    POSITIONAL, NAMED
+  }
+
+  public static final ParameterContext EMPTY_CONTEXT = new ParameterContext() {
+    @Override
+    public int addPositionalParameter(int position) throws SQLException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int addNamedParameter(int position, String bindName) throws SQLException {
+      throw new UnsupportedOperationException();
+    }
+  };
+
+  private static final List<Integer> NO_PLACEHOLDERS = Collections.emptyList();
+
+  private @Nullable BindStyle bindStyle = null;
+  private @Nullable List<Integer> placeholderPositions = null;
+  private @Nullable List<String> placeholderNames = null;
+  private @Nullable List<Integer> placeholderAtPosition = null;
+
+  /**
+   * Adds a positional parameter to this ParameterContext. Once a positional parameter have been
+   * added all subsequent parameters must be positional. Positional parameters cannot be reused, and
```

Comment:
this states "positional parameters cannot be reused", but is that actually enforced?

Reply:
Every positional parameter is a new parameter, so I would call that an intrinsic property.

Reply:
but if add were called twice with the same position, what happens?
I am not certain it is worth work to prevent, but might be good to know consequences?

Reply:
If this happens, the parser is stuck in an infinite loop, and we run out of memory :)

Reply:
I like resilient code, so I tried to add some belts, suspenders and garters. 
Positions are now only accepted in increasing order. I tried to add a few checkerframework annotations, but I think the analytical capabilites are a bit too rudimentary.

Reply:
@bokken has this been resolved ?

Reply:
yes